### DISCUSSION TITLE
T41U5XBB v209 Laser power modulation vs velocity

### DISCUSSION THREAD

**Discussion by isfl** [phil-barrett/grblHAL-teensy-4.x#94] [Q&A]
Hi there,
yesterday we did some partly successful tests with our new Teensy setup. Using IOSender. I'm attaching an image of an issue we are trying to solve. We have a particular application of laser welding (TPU backed fabrics or plain TPU film). In this application laser power modulation in corners or velocity changes is critical, as it leaves unwelded spots behind or burns through material. In the image I attach you see some areas circled in red, these areas are supposed to have continous weld geometry, but due to velocity change, some of the geometry is not welded. Same happens in corners and at entry/exit point in periodic curves (closed curves). Some of these issues can be badly mitigated by increasing overall laser power as you can see in the image, but then the weld line is still inconsistent in quality in the corners. I've been reading about a ppi based laser power modulation - but there is little info on this. Could that be the solution to the velocity change laser power issue ? Any other ideas how to improve this ? I played with a "min rpm" setting on the previous controller (8bit arduino) and the issue persisted or was creating burns instead of unwelded spots, the "min rpm" did not seem to help.

Also, we can only get relatively decent continous welds with low feed and the machine is capable of much faster speeds. If I increase feed the corner issue is much worse. Everything in the image is using F1500, F1600. I can drive it up to F12000 theoretically, but that's too scary. At least I would like to reach F3000.

I have to mention that the laser we use is analog, so using 0-10V pin to drive it. I am aware of the 10usec delay from PWM>Analog conversion (it was supposed to be around 10usec on the 8 bit Arduino).

Any help with this please ? Anyone had a similar application and worked around it ?
![IMG20231123105617](https://github.com/phil-barrett/grblHAL-teensy-4.x/assets/42996509/f58aebdc-eaf6-438f-a6af-c3b1a8de0645)

**Comment by bLouChip** [comment]
This may explain what is happening in those corners...
https://github.com/gnea/grbl/wiki/Grbl-v1.1-Laser-Mode

**Reply by isfl** [reply]
I read all this info and everything I could find on the subject. I wonder if the ppi drive mode will make a difference. The power vs velocity ramping is not very useful without some exposed parameters, for particular applications. For most cutting applications the current grbl is just fine.

**Comment by phil-barrett** [comment]
You are beyond my experience level with lasers. Have you tried a laser specific CAM/sender like Lightburn or LaserGrbl?  

I wonder what happens if you leave it in spindle mode ($32=0)?  Or, play with acceleration?  And, perhaps way into the weeds, play with junction deviation - $11 ?  There may be an interaction between laser mode and JD.

I don't think PPI is meant for this kind of application (more gray scale image rendering).

If all else fails, you might ask in the grblHAL core [discussions section](https://github.com/grblHAL/core/discussions).

What ever you come up with, I would like to know.  This is pretty interesting to me.  Please post back what you learn, even the ones that don't work!  (channeling Edison, here...)

**Comment by isfl** [comment]
Hi there,
after some tinkering with min RPM, I get reasonable welds on velocity change, and managing F2000 with decent corners. I played with the other parameters but was unable to find a sweet spot. In spindle mode, it's a disaster as expected. The previous 8 bit motion controller was not scaling the power correctly using "min RPM" and I assumed the Teensy would fail as well, but it turns out it is enough to improve things considerably. JD helps to some extent but not enough to make a difference as it has a ceiling. Overall, the Teensy made a huge difference so far. We were unable to weld curves with the 8bit controller.
![Uploading IMG20231127101940.jpg…]()

**Reply by phil-barrett** [reply]
I'm glad you are getting better results.  Thanks for posting.  What are you using for CAD and CAM?  

I wonder if the spindle linearization stuff could make a difference in correcting the actual power level.  2000 mm/min is ok but I think we should be able to do better.

**Reply by isfl** [reply]
I'm sure faster is possible, but one of the lens systems is gliding over the material, I'm a bit concerned about the surface friction. I have another non contacting lens system and that one could go much faster.

**Reply by isfl** [reply]
We use Rhino and Gcode generated by LB. We consider writing the gcode ourselves, but so far LB does a good job.

**Reply by phil-barrett** [reply]
Thanks. I didn't realize there was contact with the material.
